sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -342,7 +344,14 @@ object FileStreamSource {
     def size: Int = map.size()
   }
 
-  private[sql] trait FileStreamSourceCleaner {
+  private[sql] abstract class FileStreamSourceCleaner {
+    protected val cleanThreadPool = ThreadUtils.newDaemonCachedThreadPool(
+      "file-source-cleaner-threadpool",
+      SQLConf.get.getConf(SQLConf.FILE_SOURCE_CLEANER_NUM_THREADS)
+    )
+
+    def stop(): Unit = cleanThreadPool.shutdown()
```

Comment:
Even we just provide "best-effort" on cleaning files, we may want to at least log warning (or at least info) messages for unhandled files. Now they're just ignored silently which is not symmetric with failing to rename/delete.

As javadoc mentioned, shutdown() doesn't wait for previously submitted tasks to complete execution; graceful shutdown would be calling awaitTermination() with proper timeout, and calling shutdownNow() and logging remaining tasks.

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html#shutdown--

Reply:
+1 doing more best-effort. I think `awaitTermination` can be added with timeout but telling which file was not deleted/archived would require more effort than what it worth. At the moment don't see how a printout of `List<Runnable>` would help (not to mentioned not found such pattern in Spark).

Reply:
How end users would know which files are failed to delete/move and they need to delete/move manually? Though we say we only provide best-effort, end users should be able to deal with the missing things. It's not same with log cleaning in SHS - old log is always eligible to clean up whenever cleanup phase is executed. Here we miss once, we never get the missed files.

Reply:
Do you think it's a real life scenario that users are doing the following time-to-time?
* Stop streaming app to see logs
* Writes/updates the log parsing + file deleter/archiver app
* Parse the huge streaming logs to see missed deletes/renames
* Starts streaming app again

The main intention is to free up source side not to have horror execution time of directory listing which still stands. I would add such complexity when there is a required use-case behind.

Reply:
As far as I see even without this change we can loose file. Clean happens after we've updated the checkpoint. If app dead happens after checkpoint update but before clean then the file is lost. If we would like to give guarantees it may lead far...

Reply:
> The main intention is to free up source side not to have horror execution time of directory listing which still stands.

You may want to revisit the description of [SPARK-20568](https://issues.apache.org/jira/browse/SPARK-20568) as well as comments in the issue. 

The origin problem is "disk space", and if we fail to delete it, we would like to guide which files can be deleted manually. Reducing the cost of listing source files is one of side effects, not the main goal. Yes that brings one of major benefits though it's a side effect, but it is still not the main goal.

Reply:
Read it through, you're right.

Have you seen my comment [here](https://github.com/apache/spark/pull/26502#discussion_r346913035)? If that's true we already loose files without debug logs. If it's true, how that loss is different that this?

Reply:
Ah sorry I missed reading through the comment. That's a good point, and we may not guarantee providing the list of files if we assume "force kill". Then I'm OK to leave it as it is. 

Maybe it would better to add such in NOTE 3 (or add to NOTE 4), like below
```
The guarantee of cleaning up source files is "best-effort". Spark may not clean up some source files in some circumstances - e.g. the application doesn't shut down gracefully, too many files are waiting to clean up.
```

as NOTE 3 doesn't cover this.

Reply:
OK, then adding `awaitTermination` as we've agreed on it.